### PR TITLE
Fix multiple tooltips showing on NavigableToolbars

### PR DESCRIPTION
**Pull request by jeryj**
The 'core/block-editor/focus-toolbar' shortcut (alt + F10) selects the active navigable toolbar's first tabbable item, but when there are multiple navigable toolbars, it places focus on all of the active navigable toolbar components. This results in an odd state of two tooltips being open. This commit scaffolds the ability to disable the keyboard shortcut from applying.

Editor screen with Toggle block inserter and Paragraph tooltips open and focus on both buttons

TODO: Figure out how to identify if multiple `NavigableToolbars` are present, or at least tell the document toolbar not to show if the block toolbar is present.

## What?
Only apply the shortcut to one `NavigableToolbar` component.

## Why?
Each `NavigableToolbar` gets an `alt + F10` (or `fn + option + F10` on mac keyboards) shortcut added to it. When there are multiple `NavigableToolbar`s components on the page, they all try to focus the first tabbable element within it, displaying multiple popovers at once.

## How?
Add `useKeyboardShortcut` prop to `NavigableToolbar` and `useToolbarFocus` components so the document `HeaderToolbar` can disable itself from using the shortcut if the block toolbar is also showing. 

If there are no selected blocks OR we're not in `edit` mode, then we'll allow the keyboard shortcut for the header toolbar to be active. There may be other conditions I'm missing to identify if a focusable block toolbar is showing.

The shortcut says `Navigate to the nearest toolbar` which means to me, "Move to the toolbar closest to my current focus in the DOM," but the implementation here is, "Move to the most specific toolbar." As in, if there is a block toolbar open, move to it. If there is only one toolbar open, move to it.

## Testing Instructions
These are all documented in the e2e tests. To see them all running, add:

```
test.use( {
	launchOptions: {
		slowMo: 500,
	},
} );
``` 

to the top of `test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js` (like, line 5) and then run these tests with `npm run test:e2e:playwright -- shortcut-focus-toolbar
 --headed`

#### Block toolbar open
- Move focus into a block to display the block toolbar. 
- Press `alt + F10` (or `fn + option + F10` on mac keyboards)
- Focus should be applied to the block toolbar

#### No selected blocks
- Hide the block toolbar by moving focus to the editor page title (or anywhere else)
- Press `alt + F10` (or `fn + option + F10` on mac keyboards)
- Focus should be in the document toolbar 

#### In Select Mode
- In the top document toolbar, change to Select mode
- Place focus on a block
- Press `alt + F10` (or `fn + option + F10` on mac keyboards)
- Focus should be in the document toolbar 

Basically, try to break it using all of the different view modes (Top Toolbar on/off, Spotlight mode, etc) and with an empty default blocks selected, a block selected, multiple blocks selected, etc. Anytime a block toolbar is showing, it should receive focus with the `alt + F10`  shortcut, and anytime it's _not_ showing, focus should be in the top toolbar.

## Screenshots or screencast 
**Broken state:**
https://user-images.githubusercontent.com/967608/230490886-5849ba0e-1910-47b0-b522-750b2f99e70c.mov

### CODE REVIEW

**Comment by draganescu** [comment]
I like this solution:

- it preserves the intended behavior
- it implements a solution to a bug that comes from the limitations of our system (finding the actually "closest" toolbar)
- the opt out behaviour is easy to embrance as in the example you implemented, based on situational constraints

Let's move forward with this solution, 👏🏻

File: test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	return page.getByRole( 'button', { name: 'Paragraph', exact: true } );
+};
+
+const useSelectMode = async ( page ) => {
+	await page.keyboard.press( 'Escape' );
+};
+
+const moveToToolbarShortcut = async ( pageUtils ) => {
+	await pageUtils.pressKeys( 'alt+F10' );
+};
```
Comment (jeryj) [inline test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js]: Even though these are mostly one-liners, I found they made the test cases below much easier to read. They behave as comments when named like this.

File: test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		await expect( documentToolbarButton( page ) ).toBeFocused();
+
+		// Test: Focus block toolbar from block content when block toolbar isn't visible
+		await editor.insertBlock( { name: 'core/paragraph' } );
+		await page.keyboard.type(
+			"Focus to block toolbar when block toolbar isn't visible"
+		);
+		await moveToToolbarShortcut( pageUtils );
+		await expect( blockToolbarButton( page ) ).toBeFocused();
+		await expect( documentToolbarTooltip( page ) ).not.toBeVisible();
```
Comment (jeryj) [inline test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js]: Initially these were separate tests. I combined the tests into groups based into view and mode instead. This cuts the test execution time down by about half, (18s for running all the tests vs 9s for the current setup). This is recommended by Playwright Best Practices of "write fewer tests but longer tests": https://playwright.dev/docs/best-practices#write-fewer-tests-but-longer-tests. I think the time saved per commit is worth having more assertions per test.

File: packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ */
+import type { Editor } from './index';
+
+/**
+ * Toggles the fixed toolbar option.
+ *
+ * @param this
+ * @param isFixed Boolean value true/false for on/off.
+ */
+export async function toggleFixedToolbar( this: Editor, isFixed: boolean ) {
```
Comment (jeryj) [inline packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts]: This is taken from https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/various/links.test.js#L275-L284. If the plan is to move the older e2e tests into playwright, then this will be useful.

File: packages/block-editor/src/components/navigable-toolbar/index.js [hunk 4]
```diff
@@ -103,8 +104,14 @@ function useToolbarFocus(
...
 
+	const focusToolbarViaShortcut = () => {
+		if ( useKeyboardFocusShortcut ) {
+			focusToolbar();
+		}
+	};
+
 	// Focus on toolbar when pressing alt+F10 when the toolbar is visible.
-	useShortcut( 'core/block-editor/focus-toolbar', focusToolbar );
+	useShortcut( 'core/block-editor/focus-toolbar', focusToolbarViaShortcut );
```
Comment (jeryj) [inline packages/block-editor/src/components/navigable-toolbar/index.js:114]: I couldn't wrap `useShortcut` in an `if` statement because it would throw a React hooks error.

File: test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js (hunk 2 above)
Comment (kevin940726) [reply test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js]: Nice improvement! 💯

File: test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js (hunk 1 above)
Comment (kevin940726) [reply test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js]: It's recommended to place these inside a [POM](https://playwright.dev/docs/pom)-style util. You can see some prior examples in [other tests](https://github.com/WordPress/gutenberg/blob/71f1f6d6dbda6d197a04a17ec1f386396315d9fc/test/e2e/specs/editor/various/inserting-blocks.spec.js#L363-L375).

File: packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts (hunk 3 above)
Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts]: Maybe we should rename it to `setIsFixedToolbar` because `toggle` seems a bit weird with the `isFixed` parameter 🤔 .

File: packages/block-editor/src/components/navigable-toolbar/index.js [hunk 5]
```diff
@@ -147,6 +154,7 @@ function useToolbarFocus(
 function NavigableToolbar( {
 	children,
 	focusOnMount,
+	useKeyboardFocusShortcut = true,
```
Comment (kevin940726) [inline packages/block-editor/src/components/navigable-toolbar/index.js]: Could we rename this to `shouldUseKeyboardFocusShortcut` as seen below? `use*` variable often suggests a hook in React and that might cause some confusion. `shouuld*` also has the benefit of suggesting the variable to be a boolean too.

File: packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts (hunk 3 above)
Comment (jeryj) [reply packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts]: Yeah, that's much better to read.

File: test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js (hunk 1 above)
Comment (jeryj) [reply test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js]: Ah, nice! Thanks for the tip!

File: packages/edit-post/src/components/header/header-toolbar/index.js [hunk 6]
```diff
@@ -39,15 +40,49 @@ function HeaderToolbar() {
...
+		let isBlockWithToolbarSelected = false;
+		// Check if we have an empty block selected
+		// If the first block in a multi selection is empty, there isn't a toolbar to show
+		const selectedBlockId =
+			getSelectedBlockClientId() || getFirstMultiSelectedBlockClientId();
+		if ( selectedBlockId ) {
+			const { name, attributes = {} } = getBlock( selectedBlockId ) || {};
+			isBlockWithToolbarSelected =
+				name && ! isUnmodifiedDefaultBlock( { name, attributes } );
+		}
```
Comment (ajlende) [inline packages/edit-post/src/components/header/header-toolbar/index.js]: I'm not sure if skipping the name check will break things, but it would be nice if this could be simplified.

```suggestion
		// Check if we have an empty block selected
		// If the first block in a multi selection is empty, there isn't a toolbar to show
		const selectedBlockId = getSelectedBlockClientId() || getFirstMultiSelectedBlockClientId()
		const isBlockWithToolbarSelected = selectedBlockId
			? ! isUnmodifiedDefaultBlock( getBlock( selectedBlockId ) )
			: false
```

**Review by ajlende — COMMENTED** [review]
**[RESOLVED]**

I found a case where I still see both tooltips.

1. `esc` to enter navigation mode
2. `return` to enter edit mode
3. `alt` + `F10` shows both tooltips

https://user-images.githubusercontent.com/5129775/231522299-0111535c-4fa0-48d7-bc6d-aefb9551d2df.mov

File: packages/edit-post/src/components/header/header-toolbar/index.js (hunk 6 above)
Comment (jeryj) [reply packages/edit-post/src/components/header/header-toolbar/index.js]: I agree -- that's simpler and it works!

File: packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ */
+import type { Editor } from './index';
+
+/**
+ * Toggles the fixed toolbar option.
+ *
+ * @param this
+ * @param setIsFixedToolbar Boolean value true/false for on/off.
+ */
+export async function toggleFixedToolbar(
```
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/editor/toggle-fixed-toolbar.ts]: Oh oops, I mean the util name, not the parameter name. `isFixed` is actually a more readable parameter name too.

**Comment by ajlende** [comment]
> There may be other conditions I'm missing to identify if a focusable block toolbar is showing.

If we want these conditions to be the same, would it be better to make a hook that controls both of them based on the application state? Basically match this?

https://github.com/WordPress/gutenberg/blob/f82dcc198d52d9aaf0d150f30bcecaf052f64204/packages/block-editor/src/components/block-tools/selected-block-popover.js#L97-L110

**Comment by jeryj** [comment]
> If we want these conditions to be the same, would it be better to make a hook that controls both of them based on the application state?

[Done](https://github.com/WordPress/gutenberg/pull/49644/commits/00d3efe29eef9665d82772d03c2343d2b4d1816d)!

File: packages/block-editor/README.md [hunk 8]
```diff
@@ -880,6 +880,18 @@ _Returns_
 
 -   `any`: Returns the value defined for the setting.
 
+### useShouldContextualToolbarShow
```
Comment (kevin940726) [inline packages/block-editor/README.md]: Question: Do we want to publicize this API just now? As soon as we make it public we have to make it backward-compatible. Maybe we can keep it internal for now?

Comment (draganescu) [reply packages/block-editor/README.md]: I agree with @kevin940726 - this looks useful only for internal use for now.

Comment (draganescu) [reply packages/block-editor/README.md]: oh I saw later why this is exported, so we can use it in the post editor later. Do we really need this `useShouldContextualToolbarShow` refactoring?

File: packages/edit-post/src/components/header/header-toolbar/index.js [hunk 9]
```diff
@@ -64,9 +77,17 @@ function HeaderToolbar() {
...
 	}, [] );
+
 	const isLargeViewport = useViewportMatch( 'medium' );
 	const isWideViewport = useViewportMatch( 'wide' );
-
+	const { shouldShowContextualToolbar, canFocusHiddenToolbar } =
+		useShouldContextualToolbarShow( selectedBlockId );
+	// If there's a block toolbar to be focused, disable the focus shortcut for the document toolbar.
+	// There's a fixed block toolbar when the fixed toolbar option is enabled or when the browser width is less than the large viewport.
+	const blockToolbarCanBeFocused =
```
Comment (draganescu) [inline packages/edit-post/src/components/header/header-toolbar/index.js:90]: Is there a similar thing going on in `edit-site` too? What is the advantage of using `useShouldContextualToolbarShow` here instead of leaving all the conditions inlined where they were and just looking at `( ( hasFixedToolbar || ! isLargeViewport ) && selectedBlockId )` here?

**Review by draganescu — DISMISSED** [review]
Looks great 👏🏻 Not sure about the refactoring and also I think we need to implement the idea in the site editor and the widgets editor too?

File: packages/edit-post/src/components/header/header-toolbar/index.js (hunk 9 above)
Comment (jeryj) [reply packages/edit-post/src/components/header/header-toolbar/index.js:90]: There was a lot of [conditions in the block toolbar](https://github.com/WordPress/gutenberg/pull/49644/files#diff-d6d7994ce45fa28128923423e19291abd7de2fad164752f0f068cb9bfa93f4a0L97-L110) and @ajlende pointed out that we might be missing some conditions. For example, my conditions didn't account for smaller viewports. The way to be sure we weren't sending focus to the top toolbar was to use all the same conditions as we are for the block toolbar. That's why we decided to consolidate them. 

All that said, I don't have a strong preference. I do like this hook though. It can be very useful for knowing if the block toolbar should be showing or if it will be focusable from a keyboard shortcut anywhere else in Gutenberg.

File: packages/block-editor/README.md (hunk 8 above)
Comment (jeryj) [reply packages/block-editor/README.md]: > Do we want to publicize this API just now?

Erring on the side of caution sounds good. What does making it internal look like? Adding `experimental` or `unstable` in front of the name?

Comment (kevin940726) [reply packages/block-editor/README.md]: At the time I left this review, it was only used in the same package. If it's also used in other packages and it's useful then I don't mind publicizing it 😅 .

Comment (kevin940726) [reply packages/block-editor/README.md]: I guess there's a fine line between whether it should be implemented as a `@wordpress/data` selector or a hook. I don't mind either but would rather keep it consistent. It's just one small thing to consider before we publicizing it :).

Comment (jeryj) [reply packages/block-editor/README.md]: @kevin940726 - I believe we tried to do use selector first, but we were blocked due to the `useViewportMatch`. @ajlende can confirm/deny this statement 😅

File: packages/edit-post/src/components/header/header-toolbar/index.js (hunk 9 above)
Comment (ajlende) [reply packages/edit-post/src/components/header/header-toolbar/index.js:90]: Having the hook means that we know it will be the same in both places which is what we want. From previous experience, if you don't do things this way in gutenberg, someone is going to update the condition in one place and miss the other place.

File: packages/block-editor/README.md (hunk 8 above)
Comment (ajlende) [reply packages/block-editor/README.md]: Yeah, the `useViewportMatch` is the problem with making it a selector. We could, in the future, do the same thing as `useViewportMatch` and update the store [when the media query changes](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/change_event). So it's a good idea to make it private.

**Review by scruffian — APPROVED** [review]
LGTM

**Comment by scruffian** [comment]
We should create issues for the follow ups here.

File: test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+		await toolbarUtils.useSelectMode();
+		await toolbarUtils.moveToToolbarShortcut();
+		await expect( toolbarUtils.documentToolbarButton ).toBeFocused();
+	} );
+
+	test.describe( 'In Top Toolbar option:', () => {
+		test.beforeEach( async ( { editor } ) => {
+			// Ensure the fixed toolbar option is on
+			await editor.setIsFixedToolbar( true );
+		} );
```
Comment (Mamaduka) [inline test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js:83]: @jeryj, the block editor settings are persisted across the tests, and this setting occasionally might leak to other tests. We should probably add cleanup `test.afterEach` and restore the toolbar setting.

I noticed while debugging #49733.

## Screenshot from the test failure
test-failed-1

Comment (jeryj) [reply test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js:83]: Nice catch! Put in a fix here: https://github.com/WordPress/gutenberg/pull/51600
